ADR/00xx-component-references.md

```diff
@@ -0,0 +1,147 @@
+  * Builds of *referent* Components never directly trigger tests, never trigger promotion, or Releases.
+* When [integration-service] notices a build of a *referer* Component (one that declares that it embeds references to others), it does testing as normal, promotes to global candidate list as normal, and it promotes to dev env as normal, and it creates Releases as normal.
+* When [integration-service] notices a build of a Component that references no other components and no other components refer to it - then it proceeds like normal, runs tests, promotes, deploys, releases, etc.
+
+### Build-service
+
+* [build-service] will propagate the digest from one Component to another as a PR – following the declared `ComponentReferences`. This is a bit like renovatebot wrapped in a controller. The [build-service] acts when it sees that a build pipelinerun has completed successfully for a *referent* image.
+  * If this is the first time it has seen this *referent* updated in this way, it will file a new pull request on the *referer* supplying the pullspec and digest of the new referent, using the same branch name as was used for the *referent* PR.
+  * If a PR on that branch already exists, update that branch with an additional commit including the new *referent* digest.
+
+The [build-service] will also update the PR it filed when the PRs that triggered it are merged or updated. It may update the description and/or it may rebase on the main branch and/or it may issue /retest. Need to figure out what we want here.
+
+### “PR dependency”
+
+TODO - describe exactly how the user specifies that one PR depends on another. PaC will have to pass this through to the PipelineRun as an annotation. For now, just assume it is possible and defined.
+
+* When [integration-service] notices a build from a PR (PR#2) that declares it depends on another PR (PR#1), it will do testing but it will also:
+  * Construct the Snapshot for the test using the image from the triggering build pipeline of PR #2 as well as image found on the latest build pipeline associated with PR#1. This lets a user test PR #2 including unmerged content from PR #1.
+  * Post the test results back on PR#1 in addition to the results it would normally post to PR#2.
+  * Post a special followup check result on PR#2 that “fails” saying, “don’t merge this. It still depends on another.”
+* When [integration-service] notices a build from a PR that declares it depends on another PR but that other PR is merged, it will do testing and it will post a followup check result that “succeeds” saying “this is mergeable, all other PRs that it depends on are merged.”
```

Comment:
When is this noticing done? Is it triggered on a push event -- detecting what PRs might have been blocked by a dependent PR merge, updating those depending PRs once they are merged?

What is the testing that is done here? If images are reused then is there any testing that is actually perfomred?

Reply:
```suggestion
* When [integration-service] notices that a PR has merged (by observing a PipelineRun for a push event), it will do testing. It will find all other open PRs that declare that they depended on the PR that just merged and it will post a followup check result that “succeeds” saying “this is mergeable, all other PRs that it depends on are merged.”
```

Reply:
Is that helpful?

Reply:
What does this mean?

> it will do testing

The push event build will be tested? The dependent PR will be tested? Integration service will test to see if there are other PRs open?

The PRs might need to be updated anyway for the new images that have been produced.

Reply:
We need to name the PRs 1 and 2 or A and B. This is getting confusing.

> When two PRs (for Component C and D respectively) have a PR Dependency between them (saying that the PR-on-D depends on the PR-on-C), then:
>
>    When integration-service sees a build stemming from the PR-on-C, it will execute testing of a snapshot containing the content from PR-on-C as normal.
>    When it sees a build stemming from PR-on-D, it will execute testing of a snapshot containing the content from PR-on-D and PR-on-C.

In this case, we have PR-on-D and PR-on-C.

I think I was originally writing that

* Imagine PR-on-C has already merged and [integration-service] sees a build from PR-on-D. It should just do stuff as normal, because PR-on-C should already be in the global candidate list. integration-service doesn't need to mark PR-on-D as "don't merge".

However, the more interesting case (not yet described) is this one:

* Imagine PR-on-C has not yet merged. [integration-service] sees a build from PR-on-D. It follows new PR dependency rules and tests the two together.
* Then, later, PR-on-C merges!! [integration-service] needs to respond to the build for PR-on-C's merge and it needs to update the github checks status for PR-on-D to say "okay to merge now".

Reply:
Ack. The not yet described case does sound more interesting than the originally intended case.

> This is getting confusing.

At least I think that there is more confusion in the verbal explanation than in the implementation.